vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/node.go

```diff
@@ -644,192 +845,17 @@ var _ = DescribeSanity("Node Service", func(sc *SanityContext) {
-				VolumeId:          vol.GetVolume().GetVolumeId(),
-				TargetPath:        sc.TargetPath + "/target",
-				StagingTargetPath: stagingPath,
-				VolumeCapability: &csi.VolumeCapability{
-					AccessType: &csi.VolumeCapability_Mount{
-						Mount: &csi.VolumeCapability_MountVolume{},
-					},
-					AccessMode: &csi.VolumeCapability_AccessMode{
-						Mode: csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
-					},
-				},
-				VolumeContext:  vol.GetVolume().GetVolumeContext(),
-				PublishContext: conpubvol.GetPublishContext(),
-				Secrets:        sc.Secrets.NodePublishVolumeSecret,
-			},
-		)
-		Expect(err).NotTo(HaveOccurred())
-		Expect(nodepubvol).NotTo(BeNil())
-
-		// NodeGetVolumeStats
-		if nodeVolumeStatsSupported {
-			By("Get node volume stats")
-			statsResp, err := c.NodeGetVolumeStats(
-				context.Background(),
-				&csi.NodeGetVolumeStatsRequest{
-					VolumeId:   vol.GetVolume().GetVolumeId(),
-					VolumePath: sc.TargetPath + "/target",
-				},
-			)
-			Expect(err).ToNot(HaveOccurred())
-			Expect(statsResp.GetUsage()).ToNot(BeNil())
-		}
-
-		// NodeUnpublishVolume
-		By("cleaning up calling nodeunpublish")
-		nodeunpubvol, err := c.NodeUnpublishVolume(
-			context.Background(),
-			&csi.NodeUnpublishVolumeRequest{
-				VolumeId:   vol.GetVolume().GetVolumeId(),
-				TargetPath: sc.TargetPath + "/target",
-			})
-		Expect(err).NotTo(HaveOccurred())
-		Expect(nodeunpubvol).NotTo(BeNil())
-
-		if nodeStageSupported {
-			By("cleaning up calling nodeunstage")
-			nodeunstagevol, err := c.NodeUnstageVolume(
-				context.Background(),
-				&csi.NodeUnstageVolumeRequest{
-					VolumeId:          vol.GetVolume().GetVolumeId(),
-					StagingTargetPath: sc.StagingPath,
-				},
-			)
-			Expect(err).NotTo(HaveOccurred())
-			Expect(nodeunstagevol).NotTo(BeNil())
+		By("runControllerTest")
+		runControllerTest(sc, c, s, cl, controllerPublishSupported, nodeStageSupported, nodeVolumeStatsSupported, 1)
+	})
+	It("should be idempotent", func() {
+		if !providesControllerService {
+			Skip("Controller Service not provided: CreateVolume not supported")
 		}
-
-		if controllerPublishSupported {
-			By("cleaning up calling controllerunpublishing")
-
-			controllerunpubvol, err := s.ControllerUnpublishVolume(
-				context.Background(),
-				&csi.ControllerUnpublishVolumeRequest{
-					VolumeId: vol.GetVolume().GetVolumeId(),
-					NodeId:   ni.GetNodeId(),
-					Secrets:  sc.Secrets.ControllerUnpublishVolumeSecret,
-				},
-			)
-			Expect(err).NotTo(HaveOccurred())
-			Expect(controllerunpubvol).NotTo(BeNil())
+		if sc.Config.IdempotentCount < 2 {
+			Skip("IdempotentCount is less than 2")
 		}
-
-		By("cleaning up deleting the volume")
-
-		_, err = s.DeleteVolume(
-			context.Background(),
-			&csi.DeleteVolumeRequest{
-				VolumeId: vol.GetVolume().GetVolumeId(),
-				Secrets:  sc.Secrets.DeleteVolumeSecret,
-			},
-		)
-		Expect(err).NotTo(HaveOccurred())
+		By("runControllerTest with Idempotent count")
```

Comment:
I still think this should be a separate "It" test because then a test failure in this part is obviously related to idempotency ("normal" test passes, "test with idempotency" doesn't).

Reply:
this code is in separate It() block now, is it diff-display that is not letting this to show clearly? or did I do it not right... I can use another By() line inside It() right?

Reply:
You are right, it's fine. I just missed the previous hunk in the diff.